cli/cmd/uninstall.go

```diff
@@ -0,0 +1,304 @@
+
+This command provides all Kubernetes namespace-scoped and cluster-scoped resources (e.g services, deployments, RBACs, etc.) necessary to uninstall Linkerd.`,
+		Example: ` linkerd uninstall | kubectl delete -f -`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return uninstallRunE()
+		},
+	}
+
+	return cmd
+}
+
+func uninstallRunE() error {
+	k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+	if err != nil {
+		return err
+	}
+
+	files, err := fetchKubernetesResources(k8sAPI)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		fmt.Print(fmt.Sprintf("%s%s", file, YamlSep))
+	}
+	return nil
+}
+
+func fetchKubernetesResources(k *k8s.KubernetesAPI) ([]string, error) {
+	options := metav1.ListOptions{
+		LabelSelector: k8s.ControllerNSLabel,
+	}
+
+	clusterRoles, err := fetchClusterRoles(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch ClusterRole resources:%v", err)
+	}
+
+	clusterRoleBindings, err := fetchClusterRoleBindings(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch ClusterRoleBinding resources:%v", err)
+	}
+
+	crds, err := fetchCustomResourceDefinitions(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch CustomResourceDefinition resources:%v", err)
+	}
+
+	roleBindings, err := fetchKubeSystemRoleBindings(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch RoleBindings from kube-system namespace:%v", err)
+	}
+
+	apiCRDs, err := fetchAPIRegistrationResources(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch APIService CRDs:%v", err)
+	}
+
+	psps, err := fetchPodSecurityPolicy(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch PodSecurityPolicy resources:%v", err)
+	}
+
+	mutatinghooks, err := fetchMutatingWebhooksConfiguration(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch MutatingWebhookConfigurations:%v", err)
+	}
+
+	validationhooks, err := fetchValidatingWebhooksConfiguration(k, options)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch ValidatingWebhookConfiguration:%v", err)
+	}
+
+	namespace, err := fetchNamespaceResource(k)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch Namespace %s:%v", controlPlaneNamespace, err)
+	}
+
+	aggregatedResources := []string{
+		clusterRoles,
+		clusterRoleBindings,
+		roleBindings,
+		crds,
+		apiCRDs,
+		psps,
+		mutatinghooks,
+		validationhooks,
+		namespace,
+	}
+
+	return generateNonEmptyResourceFiles(aggregatedResources), nil
+}
+
+func generateNonEmptyResourceFiles(resources []string) []string {
+	nonEmptyResources := make([]string, 0)
+	for _, resource := range resources {
+		if resource != "" {
+			nonEmptyResources = append(nonEmptyResources, resource)
+		}
+	}
+	return nonEmptyResources
+}
+
+func fetchClusterRoles(k *k8s.KubernetesAPI, options metav1.ListOptions) (string, error) {
```

Comment:
I notice that we have a mix of list and non-list type resources in the output, which probably works, but is a bit confusing.  I'd prefer to just have a sequence of non-list resources.  I think we can accomplish that by having all of these `fetch...` functions return a slice of `ObjectReference` or something like that.

Reply:
@adleong great suggestion will get it underway! Since we only need kinds, name and version, I thought maybe it's better to have our own wrapper objects around the metadata we want to marshal? In that case, all functions can return a slice of "wrapperStruct".